planner/core/planbuilder_test.go

```diff
@@ -172,3 +185,317 @@ func (s *testPlanBuilderSuite) TestDisableFold(c *C) {
+	return checkDeepCloned(reflect.ValueOf(p), reflect.ValueOf(cloned), typeName(reflect.TypeOf(p)), whiteList, nil)
+}
+
+// checkDeepCloned is used to check if v2 is deep cloned from v1.
+// It's modified from reflect.deepValueEqual.
+func checkDeepCloned(v1, v2 reflect.Value, path string, whiteList []string, visited map[visit]bool) error {
+	if !v1.IsValid() || !v2.IsValid() {
+		if v1.IsValid() != v2.IsValid() {
+			return errors.Errorf("invalid")
+		}
+		return nil
+	}
+	if v1.Type() != v2.Type() {
+		return errors.Errorf("different type %v, %v, path %v", v1.Type(), v2.Type(), path)
+	}
+
+	if visited == nil {
+		visited = make(map[visit]bool)
+	}
+	hard := func(k reflect.Kind) bool {
+		switch k {
+		case reflect.Map, reflect.Slice, reflect.Ptr, reflect.Interface:
+			return true
+		}
+		return false
+	}
+	if v1.CanAddr() && v2.CanAddr() && hard(v1.Kind()) {
+		addr1 := unsafe.Pointer(v1.UnsafeAddr())
+		addr2 := unsafe.Pointer(v2.UnsafeAddr())
+		if uintptr(addr1) > uintptr(addr2) {
+			addr1, addr2 = addr2, addr1
+		}
+		typ := v1.Type()
+		v := visit{addr1, addr2, typ}
+		if visited[v] {
+			return nil
+		}
+		visited[v] = true
+	}
+
+	switch v1.Kind() {
+	case reflect.Array:
+		for i := 0; i < v1.Len(); i++ {
+			if err := checkDeepCloned(v1.Index(i), v2.Index(i), fmt.Sprintf("%v[%v]", path, i), whiteList, visited); err != nil {
+				return err
+			}
+		}
+	case reflect.Slice:
+		if v1.IsNil() && v2.IsNil() || (v1.Len() == 0 && v2.Len() == 0) {
+			return nil
+		}
+		if v1.Len() != v2.Len() {
+			return errors.Errorf("different slice lengths, len %v, %v, path %v", v1.Len(), v2.Len(), path)
+		}
+		if v1.IsNil() != v2.IsNil() {
+			if v1.Len() == 0 && v2.Len() == 0 {
+				return nil // nil and an empty slice are accepted
+			}
+			return errors.Errorf("different slices nil %v, %v, path %v", v1.IsNil(), v2.IsNil(), path)
+		}
+		if v1.Pointer() == v2.Pointer() {
+			return errors.Errorf("invalid slice pointers, path %v", path)
+		}
+		for i := 0; i < v1.Len(); i++ {
+			if err := checkDeepCloned(v1.Index(i), v2.Index(i), fmt.Sprintf("%v[%v]", path, i), whiteList, visited); err != nil {
+				return err
+			}
+		}
+	case reflect.Interface:
+		if v1.IsNil() && v2.IsNil() {
+			return nil
+		}
+		if v1.IsNil() != v2.IsNil() {
+			return errors.Errorf("invalid interfaces, path %v", path)
+		}
+		return checkDeepCloned(v1.Elem(), v2.Elem(), path, whiteList, visited)
+	case reflect.Ptr:
+		if v1.IsNil() && v2.IsNil() {
+			return nil
+		}
+		if v1.Pointer() == v2.Pointer() {
+			typeName := v1.Type().String()
+			inWhileList := false
+			for _, whiteName := range whiteList {
+				if whiteName == typeName {
+					inWhileList = true
+					break
+				}
+			}
+			if inWhileList {
+				return nil
+			}
+			return errors.Errorf("same pointer, path %v", path)
+		}
+		return checkDeepCloned(v1.Elem(), v2.Elem(), path, whiteList, visited)
+	case reflect.Struct:
+		for i, n := 0, v1.NumField(); i < n; i++ {
```

Comment:
Should we check `v1.NumField() == v2.NumField()` first?

Reply:
We have already checked their types `v1.Type() != v2.Type()`.